### PR TITLE
Refactor LUFactor into LU, update Solve, and add specific SolveLU rou…

### PR DESCRIPTION
**Pull request by btracey**
…tines

Updates Issue 169

### CODE REVIEW

**Comment by btracey** [comment]
Do you think LU.Factorize should return a boolean? Seems the proper time to handle the error is at the Solve function. One can use the factorization in the other functions successfully (det, etc.) even if the matrix is singular.

**Comment by btracey** [comment]
Thinking about it, we need a better solution to Permutation. It's confusing that it works backward on the indices. I'm not sure what the best way to resolve this issue is. Invert the pivots as returned by LU? Have Permutation and PermutationInv?

**Comment by kortschak** [comment]
>Do you think LU.Factorize should return a boolean? Seems the proper time to handle the error is at the Solve function. One can use the factorization in the other functions successfully (det, etc.) even if the matrix is singular.

If we save effort by knowing early that the factorisation was of a singular matrix then yes, otherwise no.

>Thinking about it, we need a better solution to Permutation. It's confusing that it works backward on the indices.

Is this just a consequence of the way LAPACK represents permutation matrices?

>I'm not sure what the best way to resolve this issue is. Invert the pivots as returned by LU? Have Permutation and PermutationInv?

What are you suggesting is the difference in behaviour between these two?

File: mat64/lu.go [hunk 1]
```diff
@@ -6,207 +6,189 @@
...
-	l := NewDense(m, n, nil)
-	for i := 0; i < m; i++ {
-		for j := 0; j < n; j++ {
-			if i > j {
-				l.set(i, j, lu.at(i, j))
-			} else if i == j {
-				l.set(i, j, 1)
-			}
+// ExtractL extracts the lower triangular matrix from an LU factorization.
+func (t *TriDense) ExtractL(lu *LU) {
```
Comment (kortschak) [inline mat64/lu.go]: Bikeshedding: `LFrom`.

File: mat64/lu.go [hunk 2]
```diff
@@ -6,207 +6,189 @@
...
-	lu := f.LU
-	m, n := lu.Dims()
-	u := NewDense(m, n, nil)
-	for i := 0; i < m; i++ {
-		for j := 0; j < n; j++ {
-			if i <= j {
-				u.set(i, j, lu.at(i, j))
-			}
+// ExtractU extracts the upper triangular matrix from an LU factorization.
+func (t *TriDense) ExtractU(lu *LU) {
```
Comment (kortschak) [inline mat64/lu.go]: Bikeshedding: `UFrom`.

File: mat64/lu_test.go [hunk 3]
```diff
@@ -5,88 +5,108 @@
...
+		b := NewDense(n, bc, nil)
+		for i := 0; i < n; i++ {
+			for j := 0; j < bc; j++ {
+				b.Set(i, j, rand.NormFloat64())
+			}
 		}
-		u := lf.U()
-		if t.u != nil {
-			c.Check(u.Equals(t.u), check.Equals, true)
+		lu := &LU{}
```
Comment (kortschak) [inline mat64/lu_test.go]: var lu LU

File: mat64/matrix.go [hunk 4]
```diff
@@ -359,7 +361,9 @@ func Det(a Matrix) float64 {
 	if a, ok := a.(Deter); ok {
 		return a.Det()
 	}
-	return LU(DenseCopyOf(a)).Det()
+	lu := &LU{}
```
Comment (kortschak) [inline mat64/matrix.go]: var lu LU (previous comment was supposed to be here - lu is not used as a non-receiver parameter).

File: mat64/matrix.go [hunk 5]
```diff
@@ -379,11 +383,14 @@ func Inverse(a Matrix) (*Dense, error) {
 func Solve(a, b Matrix) (x *Dense, err error) {
 	switch m, n := a.Dims(); {
 	case m == n:
-		lu := LU(DenseCopyOf(a))
-		if lu.IsSingular() {
+		lu := &LU{}
```
Comment (kortschak) [inline mat64/matrix.go]: var lu LU - not used as non-receiver parameter.

**Comment by btracey** [comment]
>>Do you think LU.Factorize should return a boolean? Seems the proper time to handle the error is at the Solve function. One can use the factorization in the other functions successfully (det, etc.) even if the matrix is singular.

>If we save effort by knowing early that the factorisation was of a singular matrix then yes, otherwise no.

I don't think we save any work. As the PR stands right now, we can save the work of computing the determinant. It's generally trivial, as it is O(n) relative to the O(n^3) of the full operation. If we allow the user to modify LU in any way (such as rank-one updates), we will have to add the determinant computation back in anyway. Furthermore, if we adopt the lapack/matlab/scipy suggestion of checking for the condition number, then we'll have to do that in the solve anyway.

This also argues that solves should return an error. We can have ErrSingular which is when the matrix is exactly singular (U(i,i) == 0), and ErrNearSingular which provide the condition number warning. 

For reference
Scipy discussion: https://github.com/numpy/numpy/issues/3755
Julia Discussion: https://github.com/JuliaLang/julia/issues/4286
Lapack function for condition number given LU factorization: http://www.netlib.org/lapack/explore-html/db/de4/dgecon_8f.html#a188b8d30443d14b1a3f7f8331d87ae60

I don't see an equivalent for QR; it looks like lapack assumes least-squares solutions are always okay.

>>Thinking about it, we need a better solution to Permutation. It's confusing that it works backward on the indices.

>Is this just a consequence of the way LAPACK represents permutation matrices?

No, it's that when you form the factorization rows can swapped in a certain order. In order to form the correct P you need to do the "unswapping" step, which is running the indexes backward.

>>I'm not sure what the best way to resolve this issue is. Invert the pivots as returned by LU? Have Permutation and PermutationInv?

>What are you suggesting is the difference in behaviour between these two?

Basically, one way is the "forward algorithm" where you loop over the indices and swap the row. There is then the inverse of that, where you swap looping over the indices backward. The documentation seems clearest in the forward case "P is a matrix that will swap row 0 with pivot[0], then row 1 with pivot[1], etc.". If that is indeed the behavior, then lu.Pivot() should return lu.pivot reversed.

File: mat64/lu.go (hunk 1 above)
Comment (btracey) [inline mat64/lu.go]: Done.

File: mat64/lu.go (hunk 2 above)
Comment (btracey) [inline mat64/lu.go]: Done.

File: mat64/matrix.go (hunk 5 above)
Comment (btracey) [inline mat64/matrix.go]: Done.

File: mat64/matrix.go (hunk 4 above)
Comment (btracey) [inline mat64/matrix.go]: Done.

**Comment by kortschak** [comment]
>I don't think we save any work. As the PR stands right now, we can save the work of computing the determinant. It's generally trivial, as it is O(n) relative to the O(n^3) of the full operation. If we allow the user to modify LU in any way (such as rank-one updates), we will have to add the determinant computation back in anyway. Furthermore, if we adopt the lapack/matlab/scipy suggestion of checking for the condition number, then we'll have to do that in the solve anyway.

>This also argues that solves should return an error. We can have ErrSingular which is when the matrix is exactly singular (U(i,i) == 0), and ErrNearSingular which provide the condition number warning. 

Agreed. Make a type Condition float64 that satisfies error and return that. This allows both cases to be encapsulated. Methods can be added for convenience: IsSingular, e.g.

>>Is this just a consequence of the way LAPACK represents permutation matrices?

>No, it's that when you form the factorization rows can swapped in a certain order. In order to form the correct P you need to do the "unswapping" step, which is running the indexes backward.

This is essentially what I mean.

>Basically, one way is the "forward algorithm" where you loop over the indices and swap the row. There is then the inverse of that, where you swap looping over the indices backward. The documentation seems clearest in the forward case "P is a matrix that will swap row 0 with pivot[0], then row 1 with pivot[1], etc.". If that is indeed the behavior, then lu.Pivot() should return lu.pivot reversed.

Making the pivot the right order for iterating forward makes sense and providing good documentation. Adding API should be a last resort here.

File: mat64/lu.go [hunk 6]
```diff
@@ -6,207 +6,189 @@
...
-	}
-	l := NewDense(m, n, nil)
-	for i := 0; i < m; i++ {
-		for j := 0; j < n; j++ {
-			if i > j {
-				l.set(i, j, lu.at(i, j))
-			} else if i == j {
-				l.set(i, j, 1)
-			}
+// ExtractL extracts the lower triangular matrix from an LU factorization.
```
Comment (kortschak) [inline mat64/lu.go]: Update doc, and below.

**Comment by btracey** [comment]
The change proposed for Permutation is wrong. Just inverting the errors doesn't fix it. I instead changed the API to be more general and as a bonus, also correct.

I didn't fix the Solve errors. The tests need big changes. Rather than change everything twice I'll just clean up Solve when it becomes a method on Dense.

I put the Condition documentation with the type so that the documentation could be shared across all of the functions. If QR finds the matrix is exactly singular, the decomposition must stop, so we cannot guarantee that a solve is always finished. We can possibly guarantee that the solve will always stop if Condition == inf, but the current documentation is a hedge until we're sure. I don't think it's good policy to stop if the condition number is large and finite, as in many cases things turn out okay anyway. Proper error handling would check that things are okay based on whatever measure the user thinks is important.

**Comment by btracey** [comment]
Sorry, Permute has some issues, fixing now.

**Comment by btracey** [comment]
Done.

**Comment by btracey** [comment]
Fixes some documentation problems. Also allows us to see if the build is actually fixed.

**Comment by btracey** [comment]
Oh darn. Still bad stack pointer.

**Comment by btracey** [comment]
Oh, no, sorry, that's with beta3. My bad.

**Comment by kortschak** [comment]
You need to rebase to see the effect of master on the push.

**Comment by btracey** [comment]
Yea.

Woo! Checks pass!

**Comment by kortschak** [comment]
The seem more complicated a type than necessary. A single slice with swap positions should be enough - following the cycle is not difficult. I think we had code in at some stage that did this and if not, there certainly is some in biogo.matrix that you can use.

**Comment by btracey** [comment]
The problem with Permute is that you can't just invert the order of the slice, that's not correct. You have to swap row n-1 first, then row n-2 (not 0 first then 1). We could have a "backward" boolean to run in the reverse direction. This allows for an arbitrary number of swaps (which is both good and bad).

**Comment by btracey** [comment]
One could also just run the effective swap algorithm in lu and see where the 1s end up, then reconstruct the necessary pivots.

**Comment by kortschak** [comment]
This last feels better.

**Comment by btracey** [comment]
Algorithm updated. PTAL.

**Comment by btracey** [comment]
@kortschak is this better?

**Comment by kortschak** [comment]
Absolutely. I sent an LGTM by mail this morning, but it obviously didn't get through.
